### PR TITLE
HIR-92. Refactoring of the availability service

### PR DESCRIPTION
**Pull request by oleg1std**
Ef core updated to 5.0.0-rc.1.20451.13
Added availability search changes:
- 1 request to the DB instead of several
- a new grouping of rooms ( by occupation request)
- a new grouping of available rates

### CODE REVIEW

File: HappyTravel.Hiroshima.Common/Infrastructure/Extensions/JsonDocumentExtensions.cs [hunk 1]
```diff
@@ -27,6 +27,12 @@ public static TResult GetValue<TResult>(this JsonDocument jsonDocument)
         }
 
 
+        public static bool NotEmpty(this JsonDocument jsonDocument)
```
Comment (kirillta) [inline HappyTravel.Hiroshima.Common/Infrastructure/Extensions/JsonDocumentExtensions.cs:30]: IsNotEmpty

File: HappyTravel.Hiroshima.Common/Infrastructure/Utilities/DateRange.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace HappyTravel.Hiroshima.Common.Infrastructure.Utilities
+{
+    public static class DateRange
+    {
+        public static bool Intersects(DateTime firstStart, DateTime firstEnd, DateTime secondStart, DateTime secondEnd)
```
Comment (kirillta) [inline HappyTravel.Hiroshima.Common/Infrastructure/Utilities/DateRange.cs:7]: IsIntersects

File: HappyTravel.Hiroshima.Common/Infrastructure/Utilities/ListHelper.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            {
+                var listOfItemsExceptFirst = listOfItems.Skip(1);
+                foreach (var item in listOfItems.First().Where(item => !selectedItems.Contains(item)))
+                {
+                    var concatenatedItems = selectedItems.Concat(new List<T> {item});
+                    foreach (var combination in GetCombinations(listOfItemsExceptFirst, concatenatedItems.ToList()))
+                        yield return combination;
+                }
+            }
+            else
```
Comment (kirillta) [inline HappyTravel.Hiroshima.Common/Infrastructure/Utilities/ListHelper.cs:40]: Might be reversed, I guess

File: HappyTravel.Hiroshima.Common/Infrastructure/Utilities/DateRange.cs (hunk 2 above)
Comment (hasmikharutyunyan) [reply HappyTravel.Hiroshima.Common/Infrastructure/Utilities/DateRange.cs:7]: maybe `Are`?

**Review by AlexeyShok — COMMENTED** [review]
New questions not found.

File: HappyTravel.Hiroshima.Common/Infrastructure/Utilities/DateRange.cs (hunk 2 above)
Comment (kirillta) [reply HappyTravel.Hiroshima.Common/Infrastructure/Utilities/DateRange.cs:7]: then `AreIntersect`

File: HappyTravel.Hiroshima.Common/Infrastructure/Utilities/ListHelper.cs (hunk 3 above)
Comment (oleg1std) [reply HappyTravel.Hiroshima.Common/Infrastructure/Utilities/ListHelper.cs:40]: In this case, the code after yield will be executed and it causes System.InvalidOperationException.

File: HappyTravel.Hiroshima.Common/Models/Seasons/SeasonRange.cs [hunk 4]
```diff
@@ -13,5 +14,8 @@ public class SeasonRange
         public DateTime EndDate { get; set; }
         
         public Season Season { get; set; }
+        
+        
+        public bool Intersects(SeasonRange range) => DateRange.AreIntersect(StartDate, EndDate, range.StartDate, range.EndDate);
```
Comment (kirillta) [inline HappyTravel.Hiroshima.Common/Models/Seasons/SeasonRange.cs:19]: Still `Intersects`

File: HappyTravel.Hiroshima.DirectContracts/Services/Availability/PaymentDetailsService.cs [hunk 5]
```diff
@@ -92,41 +74,42 @@ private List<decimal> GetDailyPrices(List<SeasonPriceDetails> seasonPrices)
...
                 {
-                    var (price, details) = AddDiscount(fromDate);
-                    if (!string.IsNullOrEmpty(details))
-                        priceDetails.Add(details);
-                    
-                    seasonPriceWithDiscount += price;
+                    var (priceWithDiscount, discountPercent, priceRemark) = ApplyDiscount(fromDate);
+                    if (!string.IsNullOrEmpty(priceRemark))
+                        priceRemarks.Add(priceRemark);
+                    appliedDiscounts.Add(discountPercent);
```
Comment (kirillta) [inline HappyTravel.Hiroshima.DirectContracts/Services/Availability/PaymentDetailsService.cs:88]: A space needed

**Review by AlexeyShok — APPROVED** [review]

**Pull request merged**